core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs

```diff
@@ -0,0 +1,249 @@
+        }
+    }
+
+    fn get_enumeration_index(&mut self, key: &zksync_types::StorageKey) -> Option<u64> {
+        match self {
+            Self::Postgres(postgres) => postgres.get_enumeration_index(key),
+            Self::Rocksdb(rocksdb) => rocksdb.get_enumeration_index(key),
+        }
+    }
+}
+
+impl<'a> From<PostgresStorage<'a>> for PgOrRocksdbStorage<'a> {
+    fn from(value: PostgresStorage<'a>) -> Self {
+        Self::Postgres(value)
+    }
+}
+
+impl<'a> From<RocksdbStorage> for PgOrRocksdbStorage<'a> {
+    fn from(value: RocksdbStorage) -> Self {
+        Self::Rocksdb(value)
+    }
+}
+
+/// A [`ReadStorageFactory`] implementation that can produce short-lived [`ReadStorage`] handles
+/// backed by either Postgres or RocksDB (if it's caught up). Always initialized as a `Postgres`
+/// variant and is then mutated into `Rocksdb` once RocksDB cache is caught up. After which it
+/// can never revert back to `Postgres` as we assume RocksDB cannot fall behind under normal state
+/// keeper operation.
+#[derive(Debug, Clone)]
+pub struct AsyncRocksdbCache {
+    pool: ConnectionPool,
+    rocksdb_cell: Arc<OnceCell<RocksDB<StateKeeperColumnFamily>>>,
+}
+
+impl AsyncRocksdbCache {
+    /// Load latest sealed miniblock from the latest sealed L1 batch (ignores sealed miniblocks
+    /// from unsealed batches).
+    async fn load_latest_sealed_miniblock(
+        connection: &mut StorageProcessor<'_>,
+    ) -> anyhow::Result<Option<(MiniblockNumber, L1BatchNumber)>> {
+        let mut dal = connection.blocks_dal();
+        let Some(l1_batch_number) = dal.get_sealed_l1_batch_number().await? else {
+            return Ok(None);
+        };
+        let Some((_, miniblock_number)) =
+            dal.get_miniblock_range_of_l1_batch(l1_batch_number).await?
+        else {
+            return Ok(None);
+        };
+        Ok(Some((miniblock_number, l1_batch_number)))
+    }
+
+    /// Returns a [`ReadStorage`] implementation backed by Postgres
+    async fn access_storage_pg(
+        rt_handle: Handle,
+        pool: &ConnectionPool,
+    ) -> anyhow::Result<PgOrRocksdbStorage<'_>> {
+        let mut connection = pool.access_storage().await?;
+
+        let (miniblock_number, l1_batch_number) =
+            match Self::load_latest_sealed_miniblock(&mut connection).await? {
+                Some((miniblock_number, l1_batch_number)) => (miniblock_number, l1_batch_number),
+                None => {
+                    tracing::info!("Could not find latest sealed miniblock, loading from snapshot");
+                    let snapshot_recovery = connection
+                        .snapshot_recovery_dal()
+                        .get_applied_snapshot_status()
+                        .await
+                        .context("failed getting snapshot recovery info")?
+                        .ok_or_else(|| {
+                            anyhow::anyhow!("Could not find snapshot, no state available")
+                        })?;
```

Comment:
Nit: You can use `anyhow::Context::context()` on `Option`s as well, so this can be shortened to `.context("Could not find snapshot, no state available")`.

Reply:
Oh, nice, had no idea it worked like that